src/uopool/services/simulation.rs

```diff
@@ -403,6 +404,126 @@ impl<M: Middleware + 'static> UoPoolService<M> {
+        &self,
+        contract_addresses: Vec<Address>,
+        code_hashes: &mut Vec<CodeHash>,
+    ) -> Result<(), SimulateValidationError> {
+        let mut tasks: JoinSet<Option<(Address, H256)>> = JoinSet::new();
+
+        for contract_address in contract_addresses {
+            let eth_provider = self.eth_provider.clone();
+
+            tasks.spawn(async move {
+                match eth_provider.get_code(contract_address, None).await {
+                    Ok(code) => Some((contract_address, keccak256(&code).into())),
+                    Err(_) => None,
+                }
+            });
+        }
+
+        while let Some(result) = tasks.join_next().await {
+            match result {
+                Ok(Some(code_hash)) => code_hashes.push(CodeHash {
+                    address: code_hash.0,
+                    hash: code_hash.1,
+                }),
+                Ok(None) | Err(_) => {
+                    return Err(SimulateValidationError::UnknownError {
+                        error: "failed to get code hashes".to_string(),
+                    });
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn compare_code_hashes(
+        &self,
+        code_hashes: &Vec<CodeHash>,
+        prev_code_hashes: &Vec<CodeHash>,
+    ) -> bool {
+        if prev_code_hashes.len() != code_hashes.len() {
+            return false;
+        }
+
+        let code_hashes_map = code_hashes
+            .iter()
+            .map(|code_hash| (code_hash.address, code_hash.hash))
+            .collect::<HashMap<Address, H256>>();
+
+        for code_hash in prev_code_hashes {
+            if let Some(hash) = code_hashes_map.get(&code_hash.address) {
+                if hash != &code_hash.hash {
+                    return false;
+                }
+            } else {
+                return false;
+            }
+        }
+
+        true
+    }
+
+    async fn code_hashes(
+        &self,
+        user_operation: &UserOperation,
+        entry_point: &Address,
+        trace: &JsTracerFrame,
+    ) -> Result<(), SimulateValidationError> {
+        let contract_addresses = trace
+            .number_levels
+            .iter()
+            .flat_map(|level| {
+                level
+                    .contract_size
+                    .keys()
+                    .copied()
+                    .collect::<Vec<Address>>()
+            })
+            .collect::<Vec<Address>>();
+
+        let code_hashes: &mut Vec<CodeHash> = &mut vec![];
+        self.get_code_hashes(contract_addresses, code_hashes)
+            .await?;
+
+        let mempool_id = mempool_id(entry_point, &self.chain_id);
+        let user_operation_hash = user_operation.hash(entry_point, &self.chain_id);
+
+        if let Some(mempool) = self.mempools.write().get_mut(&mempool_id) {
+            match mempool.has_code_hashes(&user_operation_hash) {
+                Ok(true) => {
+                    // 2nd simulation
```

Comment:
This is an interesting way to note it as 2nd simulation.

I think it makes more sense to have an explicit compare.

Reply:
I was also thinking of wrapping user operation, but that would mean passing user operation around with additional data that is not needed most of the time (except here).

Reply:
One option is to add flag parameter to simulate function.

1st simulation is always when adding user operation to mempool.
2nd simulation is when bundling.

We can pass true/false to the function on that places.

Reply:
I don't have other ideas now. Let's solve it later when we got another good plan.